### PR TITLE
Enable short-name access to TestContext devices

### LINKED ISSUE #357: How to connect DeviceProxy to device launched via MultiDeviceTestContext
Tests that use `MultiDeviceTestContext` will sometimes pass, sometimes error out with a message objecting that environment variable `TANGO_HOST` is not defined.

If you set that environment variable, `MultiDeviceTestContext` will try to connect to a database at that location. And if you stand up a database there, it will talk to that database!

I can see in the code that the tango database is supposed to be mocked out with a text file. It seems that that mocking isn't always working.

I haven't found a way to fix the problem, but I can make the problem happen every time by putting a `time.sleep(5)` at the top of the post init callback. This makes me think that there is some kind of race condition happening.

The issue occurs regardless of whether your asynchronous context is threads or processes, but it is worse for threads.

### PR DESCRIPTION
**Pull request by ajoubertza**
### Easier client access during tests

When a device server is launched via the `(Multi)DeviceTestContext`, it is run without a Tango database.  This means that access to any of its devices must be done via a fully qualified domain name (FQDN).  I.e., we need something like `tango://172.17.0.3:48493/test/device/1#dbase=no`, instead of just the short name: `test/device/1`.  Requiring the FQDN is inconvenient.

This change stores the test device server's FQDN in an environment variable while the TextContext is active.  Then, in order to make device access with short names transparent, the `DeviceProxy`, `AttributeProxy` and `Group` classes have been modified to take this environment variable into account.  If it is set, and a short (i.e., unresolved) name is used for the proxy constructor or group add method, then the name is rewritten to a fully qualified (resolved) name.

Limitations:
- Group patterns (`*` wildcard) are not supported.
- Group device names will be FQDNs, rather than the short name that was originally supplied.
- Launching two TestContexts in the same process will not work correctly without FQDNs.  The TestContext that is started second will overwrite the environment variable set by the first.

There should not be a noticeable performance degradation for this outside of test environments. It it just a single environment variable lookup per proxy constructor or group add invocation.

This new TestContext functionality is enabled by default.  It can be disabled by setting the `enable_test_context_tango_host_override` attribute to `False` before starting the TestContext.  This may also be required if you wanted a test device to connect real device in a Tango facility.  Alternatively, the override could be left enabled, and instead specify the FQDN for the real device.

### Pick open port for DeviceTestContext early

This is a minor change to the test context behaviour to support the short-name access.  If the caller doesn't specify a port, instead of letting Tango select an open port, we get one just before starting the `(Multi)DeviceTestContext`. This way the port is known before starting, and we can use it in the new environment variable.

There is a tiny chance of race condition.  The OS could allocate the port to another process between object construction and when the Tango device is actually launched.  The risk is considered very low.  We have been using this approach for the `test_event.py` tests for years and haven't seen failures in CI due to that phenomenon.  Running in a Docker or dedicated CI build machine reduces the likelihood a new ports being allocated to external processes.

----------------------
Closes #357

### CODE REVIEW

**Comment by ajoubertza** [comment]
@DrewDevereux Please have a look at this PR, if possible.

**Comment by DrewDevereux** [comment]
Hi @ajoubertza, thanks for inviting me to have a look. I haven't had time yet to look into the code, but I want to share a thought on the ephemeral port race condition.
Firstly, my repo's test harness uses `python-xdist` to distribute tests across available cores (`--dist loadscope --numprocesses auto`). This gives us a huge speed-up in unit testing, but the risk of seeing the race condition is not low 
 in this setup. In fact, we see it often enough that we have had to turn off test distribution in the CI pipeline.
A little while ago I was looking at section 6.6 of http://omniorb.sourceforge.net/omni42/omniORB/omniORB006.html, and noted that if we leave the port empty, or provide a port range, then omniORB will choose an ephemeral port itself. I wonder if it would be possible to let omniORB choose the port, and then find out what port it chose?

**Comment by ajoubertza** [comment]
Thanks for that useful observation @DrewDevereux.  The default behaviour of the TestContext was to use let OmniORB select an ephemeral port.  I'll see if there is a way to stick with that.  There are a couple of issues:
- Make sure the environment variable is set in both parent and sub process (if using `process=True`), when the port is only known after the sub process is started.  I think this is straight forward - the test context does have hooks at the right places.
- How to have the environment variable set before, or at the start of, the device server's `init_device` method, when we only get the port number after server initialisation is complete.  Users might create proxies in their `init_device` methods.

**Comment by ajoubertza** [comment]
Beside the ephemeral ports issue, there is something fascinating about the behaviour of `test_multi_short_name_access_fails_if_override_disabled` on Travis.  It is failing unexpectedly (not raising the expected exception).  On Python 2.7 we get a `RuntimeError`, instead of `DevFailed` - not sure why.  More interesting are the Python 3 test results - no exception is raised when creating the `DeviceProxy` using the short-name access.  The link to the device is successful because the proxy is created from the same process as the device server, so we can do in-process comms without contacting the Tango database.  After some digging, this makes sense since the tests are not setting `process=True`.  However, in my Docker test environment the behaviour differs - the `DeviceProxy` is still trying to access the Tango database, even with `process=False`.  I don't know why the behaviour differs in the Travis Ubuntu VM versus the Conda Docker image I am using.  Maybe something network related?

The bottom line is that with `process=False` for the TestContext, this whole PR might not be required on _some_ environments.  The goal is now to get consistent behaviour across both environments (Travis VM and my Docker).
